### PR TITLE
perf: specialize persistent BatchAttention for equal KV strides

### PR DESCRIPTION
**Pull request by saltyminty**
## 📌 Description

Persistent BatchAttention computes separate K/V offsets even when their strides match. Add an equal-stride specialization that shares offset calculation and prefetch while preserving separate K/V pointers and independent NVFP4 scale addressing.

The wrapper and default AOT build only the equal-stride primary; unequal inputs lazily load a separate specialization. Selection checks all three original data strides on every run. Public/custom generators retain their signatures, URIs, and runtime compatibility. Reuse the existing prefill lazy loader and expose the same prewarm API for graph capture.

**Performance:** baseline 35a3e9730f2f94bf2e890e2888d5c82cecd12d9f, CUDA 13.0.88, PyTorch 2.13.0+cu130, release SM100a/SM90a. Across 30 B200 primary cases, geometric-mean speedup is **1.0456x**, with conservative noise-expanded 95% interval **[1.0207x, 1.0707x]**. This remains inconclusive against a preselected 3% improvement threshold.

Representative direct-run medians in microseconds, BF16 NHD, equal strides, page size 16, 32 query / 8 KV heads, head dimension 128, causal:

| Request groups (count, KV length, query length) | B200 before → after | H100 before → after |
| --- | --- | --- |
| (1, 8192, 1) | 23.936 → 23.872 | 30.544 → 31.248 |
| (64, 8192, 1) | 330.672 → 322.400 | 726.176 → 725.904 |
| (2, 4099, 129) | 98.720 → 96.320 | 112.896 → 113.312 |
| (64, 8192, 1) + (1, 8192, 1024) | 730.048 → 693.200 | 1069.744 → 1058.528 |
| (100, 1024, 1) + (8, 8192, 17) | 215.792 → 204.704 | 273.856 → 270.096 |

Five alternating baseline/candidate rounds, 100 synchronized CUDA-event samples per case after warmup; medians above summarize five round medians. Direct timing includes launch gaps. Separate B200 graph replay: batched decode 326.528 → 317.744 us; mixed 726.112 → 687.648 us. The primary matrix crosses these five groups with page sizes 1/16 and head configurations (32,8,128), (64,8,128), (64,4,64).

Default 36-module shared-library bytes decrease: B200 22,135,104 → 21,764,128 (-1.68%); H100 19,816,768 → 19,535,904 (-1.42%). Median cold compile/link time across three builds is 280.616 → 282.883 s and 376.210 → 368.075 s. Using both stride modes still retains both libraries; representative cold independent compilation costs 6.998 s on B200 / 9.331 s on H100.

<details>
<summary>Timing reproduction recipe</summary>

Run this extracted probe in separate source-installed baseline and PR environments on the same GPU, with FLASHINFER_CUDA_ARCH_LIST=10.0a (B200) or 9.0a (H100). Use separate JIT caches, warm them first, and alternate five baseline/candidate runs. Set groups to each row above and vary page/hq/hk/dim for the primary matrix. This recipe is not an additional measurement.

```python
import itertools
import statistics
import torch
from flashinfer.attention import BatchAttention

torch.manual_seed(1715)
groups = [(64, 8192, 1)]
page, hq, hk, dim = 16, 32, 8, 128
lengths = [nk for count, nk, nq in groups for _ in range(count)]
queries = [nq for count, nk, nq in groups for _ in range(count)]
pages = [(n + page - 1) // page for n in lengths]
def pointer(sizes):
    return torch.tensor([0] + list(itertools.accumulate(sizes)), dtype=torch.int32)
q = torch.randn(sum(queries), hq, dim, dtype=torch.bfloat16, device="cuda")
shape = (sum(pages), page, hk, dim)
k = torch.randn(shape, dtype=q.dtype, device="cuda")
v = torch.randn_like(k)
out = torch.empty_like(q)
lse = torch.empty(q.shape[:2], dtype=torch.float32, device="cuda")
indices = torch.arange(sum(pages), dtype=torch.int32, device="cuda")
wrapper = BatchAttention(kv_layout="NHD", device="cuda:0")
for _ in range(6):
    wrapper.plan(pointer(queries), pointer(pages), indices,
                 torch.tensor(lengths, dtype=torch.int32), hq, hk, dim, dim, page,
                 causal=True, q_data_type=q.dtype, kv_data_type=q.dtype)
    torch.cuda.synchronize()
def run():
    wrapper.run(q, (k, v), out=out, lse=lse)
for _ in range(40):
    run()
torch.cuda.synchronize()
starts = [torch.cuda.Event(enable_timing=True) for _ in range(100)]
ends = [torch.cuda.Event(enable_timing=True) for _ in range(100)]
samples = []
for start, end in zip(starts, ends):
    start.record()
    run()
    end.record()
    end.synchronize()
    samples.append(start.elapsed_time(end) * 1000)
print(statistics.median(samples), "us")
```

</details>

## 🔍 Related Issues

Follow-up to #4736, applying its lazy specialization approach to persistent BatchAttention.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

Changed-file hooks passed, including clang-format, mypy, and Ruff; the all-files command above was not run.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Targeted validation in source-installed GPU environments:

- Final stride/trace contracts: B200 23 passed; H100 19 passed with four NVFP4 cases deselected. Generator checks: 8 passed on each.
- Unchanged shared/wrapper lazy-loader tests: 25 passed. Direct public/custom generators: 4 pytest cases / 12 plan-run evaluations passed on both GPUs. Prefill regression probes passed.
- Coverage includes NHD/HND, each unequal data-stride axis, FP16/BF16, all four NVFP4 data/scale stride combinations, wrapper reuse, graph capture/prewarm/replay with changed inputs, and real no-JIT provider/missing-artifact behavior.
- AOT qualification preserves exact baseline parity: 96 passes and 48 pre-existing mixed16 failures per variant/GPU. All 72 candidate libraries load; missing-independent-cache checks pass. Known failures are not counted as passes.
- Existing evidence is reused where source hashes are unchanged. The full repository suite and installed-wheel/provider metadata were not qualified; CI has not been requested.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

Unequal layouts should call prewarm_paged_kv_stride_variant("independent") after planning and before CUDA graph capture. With JIT disabled, the independent variant requires a recognized provider artifact or an already-loaded holder. Default AOT intentionally includes only the primary.

H100 shows little runtime benefit. NVFP4 diagnostic point estimates were 1.4–2.6% slower, with no repeatable slowdown above 3% under the same rule; this PR makes no NVFP4 speedup claim. B200 profiled spill instructions decreased from 14,360 to 0, while registers/thread and occupancy were essentially unchanged.

## Summary by CodeRabbit

* **New Features**
  * Batch attention supports paged K/V caches with equal or independent memory strides.
  * Runtime dispatch selects the appropriate implementation based on cache layout.
  * Added prewarming for independent-stride execution to support CUDA graph capture.
  * Improved guidance when required JIT modules are unavailable.

* **Bug Fixes**
  * Improved correctness across page-, token-, and head-strided cache layouts.

* **Tests**
  * Expanded coverage for stride handling, lazy loading, CUDA graphs, AOT builds, tracing, and custom attention modules.

### CODE REVIEW

**Comment by saltyminty** [comment]
/bot run tests/attention/test_batch_attention.py tests/attention/test_batch_attention_strides.py tests/attention/test_batch_prefill_kernels.py tests/jit/test_attention_utils.py tests/jit/test_batch_attention_generator_runtime.py tests/jit/test_batch_attention_lazy_loader.py tests/jit/test_batch_prefill_lazy_loader.py tests/jit/test_paged_kv_lazy_loader.py tests/trace/test_batch_attention_run_reference_correctness.py

**Comment by qsang-nv** [comment]
**1. `/bot run` is red on SM120.** `too many blocks in cooperative launch` in `test_batch_attention_strides`, `test_batch_attention_generator_runtime` and `test_batch_attention_run_reference_correctness` (8/8/4 failures across CUDA 13.0 and 13.4 combined, plus 4/4/2 on CUDA 12.9, for which the report could not compare against a baseline).

`TwoStageHolisticPlan` (`scheduler.cuh:1243`) plans `2 × SM_count` blocks for head dimensions below 256 and `1 ×` otherwise, without querying kernel occupancy, so the launch fails when that grid exceeds the device-wide resident-block capacity. These failures are consistent with the pre-existing SM120/121 limitation already recorded by three `@pytest.mark.xfail(get_compute_capability(...)[0] == 12, ...)` markers in `tests/attention/test_batch_attention.py`; the new and rewritten tests do not carry that handling. For the trace test this is a coverage change — it previously exercised `BatchDecodeWithPagedKVCacheWrapper` behind a broad `except: pytest.skip(...)`, and moving it to the actual persistent API brings that API's SM120/121 limitation into it.

**2. `prewarm_paged_kv_stride_variant()` does not resolve the dispatch for the calling device.** `prewarm()` (`_core.py:258-260`) only calls `spec.build_and_load()`. When at least one loadable AOT artifact for the independent module covers another visible architecture, but none covers the wrapper's device, `try_load()` returns a `_HeterogeneousAOTModule` instead of resolving or building a module for that device (`jit/core.py:528-562`), and `__getattr__` hands back a `dispatch` closure without touching any architecture-specific copy. `prewarm()` then returns successfully with `is_loaded == True`. At the first unequal-stride `run()`:

- `run()` (`_core.py:351-358`) takes the warm branch and skips `_check_not_capturing()`; dispatch falls through `_select_attribute()` → `_get_fallback_module()` → `_build_and_load_jit_fallback()` → `self.build(); self.load()` (`jit/core.py:375-405`, `564-567`). That is a FlashInfer-side compile, not driver PTX JIT, at the point prewarm exists to move work away from.
- With JIT disabled, `build_and_load()` returns from `try_load()` at `jit/core.py:337-346`, before the `FLASHINFER_DISABLE_JIT` check at line 348, so `MissingJITCacheError` surfaces from `build()` (line 570) at that same first `run()` rather than at `prewarm()`.

The primary is not exposed this way: `plan()` calls `self.module.plan(...)` (`_core.py:220`) with CUDA tensors, so `_cuda_architectures_for_call()` (`jit/env.py:320-352`) resolves the dispatch, and any fallback build, during `plan()`.

Scope of the second finding:: this comes from the shared lazy loader batch-prefill already uses, and needs an AOT provider carrying the independent module with architecture metadata. `test_lazy_no_jit_recognized_cache_provider` does not reach it — it installs through `FLASHINFER_AOT_DIR`, where `cuda_architectures is None` short-circuits at `jit/core.py:545-546` before the heterogeneous branch.

**Comment by saltyminty** [comment]
/bot run tests/attention/test_batch_attention.py tests/attention/test_batch_attention_strides.py tests/attention/test_batch_prefill_kernels.py tests/jit/test_attention_utils.py tests/jit/test_batch_attention_generator_runtime.py tests/jit/test_batch_attention_lazy_loader.py tests/jit/test_batch_prefill_lazy_loader.py tests/jit/test_paged_kv_lazy_loader.py tests/jit/test_jit_cache_providers.py tests/trace/test_batch_attention_run_reference_correctness.py

**Review by qsang-nv — APPROVED** [review]

**Comment by saltyminty** [comment]
/bot run tests/attention/test_batch_attention.py tests/attention/test_batch_attention_strides.py tests/attention/test_batch_prefill_kernels.py tests/jit/test_attention_utils.py tests/jit/test_batch_attention_generator_runtime.py tests/jit/test_batch_attention_lazy_loader.py tests/jit/test_batch_prefill_lazy_loader.py tests/jit/test_paged_kv_lazy_loader.py tests/jit/test_jit_cache_providers.py tests/trace/test_batch_attention_run_reference_correctness.py
